packages/appium/docs/en/ecosystem/build-plugins.md

```diff
@@ -2,4 +2,436 @@
+}
+
+async execute(next, driver, script, args) {
+  return await this.executeMethod(next, driver, script, args);
+}
+```
+
+We have three important components shown here which make this system work, all of which are defined
+inside the plugin class:
+
+1. The `executeMethodMap`, defined in just the same way as for drivers
+1. The implementation of the command method as defined in `executeMethodMap` (in this case,
+   `plugMeIn`)
+1. The overriding/handling of the `execute` command. Just like any plugin command handlers, the
+first two arguments are `next` and `driver`, followed by the script name and args. `BasePlugin`
+implements a helper method which we can simply call with all of these arguments.
+
+Overriding Execute Methods from drivers works as you'd expect: if your plugin defines an Execute
+Method with the same name as that of a driver, your command (in this case `plugMeIn`) will be
+called first. You can choose to run the driver's original behaviour via `next` if you want.
+
+### Update the Appium server object
+
+You probably don't normally need to update the Appium server object (which is an Express server
+having already been configured in a variety of ways). But, for example, you could add new Express
+middleware to the server to support your plugin's requirements. To update the server you must
+implement the `static async updateServer` method in your class. This method takes three parameters:
+
+* `expressApp`: the Express app object
+* `httpServer`: the Node HTTP server object
+* `cliArgs`: a map of the CLI args used to start the Appium server
+
+You can do whatever you want with them inside the `updateServer` method. You might want to
+reference how these objects are created and worked with in the BaseDriver code, so that you know
+you're not undoing or overriding anything standard and important. But if you insist, you can, with
+results you'll need to test! Warning: this should be considered an advanced feature and requires
+knowledge of Express, as well as the care not to do anything that could affect the operation of
+other parts of the Appium server!
+
+### Handle unexpected session shutdown
+
+When developing a plugin you may want to add some cleanup logic for when a session ends. You would
+naturally do this by adding a handler for `deleteSession`. This works in most cases, except when
+the session does not finish cleanly. Appium sometimes determines that a session has finished
+unexpectedly, and in these situations, Appium will look for a method called `onUnexpectedShutdown`
+in your plugin class, which will be called (passing the current session driver as the first
+parameter, and the error object representing the cause of the shutdown as the second), giving you
+an opportunity to take any steps that might be necessary to clean up from the session. For example,
+keeping in mind that the function is not `await`ed you could implement something like this:
+
+```js
+async onUnexpectedShutdown(driver, cause) {
```

Comment:
I assume other examples should also be more explicit about these methods should not be global but rather plugin methods